### PR TITLE
Allow language ID to be overridden via eglot-server-programs

### PR DESCRIPTION
**Pull request by purcell**
As discussed in #677.

### CODE REVIEW

File: eglot-tests.el [hunk 1]
```diff
@@ -1046,11 +1047,23 @@ will assume it exists."
...
+  (let ((major-mode 'foo-mode))
+    (let ((eglot-server-programs '((foo-mode . ("prog-executable")))))
+      (eglot--guessing-contact (_ _ _ _ guessed-lang)
+        (should (equal guessed-lang "foo"))))
+    (let ((eglot-server-programs '(((foo-mode . "bar") . ("prog-executable")))))
+      (eglot--guessing-contact (_ _ _ _ guessed-lang)
+        (should (equal guessed-lang "bar"))))
+    (let ((eglot-server-programs '(((baz-mode (foo-mode . "bar")) . ("prog-executable")))))
+      (eglot--guessing-contact (_ _ _ _ guessed-lang)
+        (should (equal guessed-lang "bar"))))))
```
Comment (purcell) [inline eglot-tests.el:1065]: Here I've grouped the associated tests together, rather than updating the existing tests.

File: eglot.el [hunk 2]
```diff
@@ -716,9 +723,26 @@ PRESERVE-BUFFERS as in `eglot-shutdown', which see."
...
+          (cl-return-from found (cons (string-remove-suffix "-mode" (symbol-name mode)) contact)))
+         ((and (consp spec)
+               (stringp (cdr spec))
+               (provided-mode-derived-p mode (car spec)))
+          (cl-return-from found (cons (cdr spec) contact))))))))
+
 (defun eglot--guess-contact (&optional interactive)
   "Helper for `eglot'.
-Return (MANAGED-MODE PROJECT CLASS CONTACT).  If INTERACTIVE is
+Return (MANAGED-MODE PROJECT CLASS CONTACT LANG-ID).  If INTERACTIVE is
```
Comment (purcell) [inline eglot.el:745]: Here I put the language ID last in the arg list, but I can move it to a different position if you prefer. e.g. after MANAGED-MODE.

File: eglot.el [hunk 3]
```diff
@@ -1735,11 +1759,7 @@ THINGS are either registrations or unregisterations (sic)."
   (append
    (eglot--VersionedTextDocumentIdentifier)
    (list :languageId
-	 (cond
-           ((get major-mode 'eglot-language-id))
-           ((string-match "\\(.*\\)-mode" (symbol-name major-mode))
-            (match-string 1 (symbol-name major-mode)))
-           (t "unknown"))
+	 (eglot--language-id (eglot-current-server))
```
Comment (purcell) [inline eglot.el:1762]: I don't know if this check needs to be guarded, ie. will there always be a current server? If so, how best to handle it?

**Review by purcell — COMMENTED** [review]
Added comments to call out particular points of potential discussion.

**Comment by purcell** [comment]
There may be some other default language ID overrides to add here, as indicated by the discussion in #624.

**Comment by joaotavora** [comment]
I'm having a look at this now.  Don't be suprised to see commits popping in.

File: eglot-tests.el (hunk 1 above)
Comment (joaotavora) [reply eglot-tests.el:1066]: :+1:

File: eglot.el (hunk 2 above)
Comment (joaotavora) [reply eglot.el:756]: :+1:

File: eglot.el (hunk 3 above)
Comment (joaotavora) [reply eglot.el]: The guard is using `eglot--current-server-or-lose` .  We must have a language id and having nil as the server will just blow up.  let it blow up with a better error.  But it shouldn't, the server should always be bound in the relevant operations.

**Comment by joaotavora** [comment]
@purcell, I've pushed a commit, minor refactors and recovers the `eglot-language-id` system, which we can't throw away.

**Comment by purcell** [comment]
> @purcell, I've pushed a commit, minor refactors and recovers the `eglot-language-id` system, which we can't throw away.

Nice, thanks. TIL the "thereis" keyword for `cl-loop`, thanks for rewriting that function.

I squashed that commit into mine, keeping your itemised list of changes.

**Comment by joaotavora** [comment]
> Nice, thanks. TIL the "thereis" keyword for cl-loop, thanks for rewriting that function.

Rewriting is the best code review :-) Thanks you for the PR. `thereis` goes with `never` and `always`.  You can exchange a `loop..thereis` for a funcional `some` and a `loop..never` for a `notany` and a `loop...always` for an `every`.  I choose depending
on the day of the week.

**Comment by joaotavora** [comment]
Should we mention this in the README under the quirky server section? I'd also like a review from @nemethf , if possible.

**Comment by purcell** [comment]
> Should we mention this in the README under the quirky server section?

Yes, I will add this.

**Comment by purcell** [comment]
Done. Added this:
----

Similarly, some servers require the language identifier strings they
are sent by `eglot` to match the exact strings used by VSCode. `eglot`
usually guesses these identifiers from the major mode name
(e.g. `elm-mode` → `"elm"`), but the mapping can be overridden using
the `LANGUAGE-ID` element in the syntax of `eglot-server-programs` if
necessary.

**Comment by joaotavora** [comment]
nice, good wording

**Comment by joaotavora** [comment]
Just force-pushed another version, I think this is good to go. But have a look @purcell

**Comment by purcell** [comment]
Yes, looks good to me!

**Pull request merged**